### ISSUE TITLE
WSL2: Canceling reading directory on drvfs/9p became WSL crash.

### ISSUE THREAD

**Issue by 0xbadfca11** [microsoft/WSL#4230]
* Your Windows build number:
Microsoft Windows [Version 10.0.18922.1000]

* What you're doing and what's happening:
Run ``while :; do find /mnt/c > /dev/null & sleep 1; killall find; done``.
All distributions will crash, hang, or cause I/O errors on drvfs.
Similar to #4219 but have not long file name. Running normally find is successful until the end.
```
'Virtual Machine' has encountered a fatal error.  The guest operating system reported that it failed with the following error codes: ErrorCode0: 0x9, ErrorCode1: 0x81000004132B0000, ErrorCode2: 0xFFFFFFFF818E09D9, ErrorCode3: 0xFFFF888061AA00C8, ErrorCode4: 0xFFFFC9000016BE20.  If the problem persists, contact Product Support for the guest operating system.  (Virtual machine ID A91B0DD1-74AC-5401-AC3C-EE15D6A7F89A)

Guest message:
```
```
'Virtual Machine' has encountered a fatal error.  The guest operating system reported that it failed with the following error codes: ErrorCode0: 0x0, ErrorCode1: 0x0, ErrorCode2: 0x0, ErrorCode3: 0x0, ErrorCode4: 0x0.  If the problem persists, contact Product Support for the guest operating system.  (Virtual machine ID A91B0DD1-74AC-5401-AC3C-EE15D6A7F89A)

Guest message:
<6>[    0.607748] Freeing unused kernel image memory: 1348K
<6>[    0.660749] Write protecting the kernel read-only data: 18432k
<6>[    0.662526] Freeing unused kernel image memory: 2008K
<6>[    0.663240] Freeing unused kernel image memory: 1836K
<6>[    0.663650] Run /init as init process
<6>[    0.675712] IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready
<5>[    0.760674] scsi 0:0:0:0: Direct-Access     Msft     Virtual Disk     1.0  PQ: 0 ANSI: 5
<5>[    0.763067] sd 0:0:0:0: Attached scsi generic sg0 type 0
<5>[    0.770347] sd 0:0:0:0: [sda] 536870912 512-byte logical blocks: (275 GB/256 GiB)
<5>[    0.770349] sd 0:0:0:0: [sda] 4096-byte physical blocks
<5>[    0.770770] sd 0:0:0:0: [sda] Write Protect is off
<7>[    0.770773] sd 0:0:0:0: [sda] Mode Sense: 0f 00 00 00
<5>[    0.771880] sd 0:0:0:0: [sda] Write cache: enabled, read cache: enabled, doesn't support DPO or FUA
<5>[    0.825263] sd 0:0:0:0: [sda] Attached SCSI disk
<6>[    1.266251] EXT4-fs (sda): recovery complete
<6>[    1.373614] EXT4-fs (sda): mounted filesystem with ordered data mode. Opts: discard,errors=remount-ro,data=ordered
<5>[    1.389370] random: fast init done
<6>[    1.600940] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
<5>[   10.975340] random: crng init done
<1>[   22.252360] general protection fault: 0000 [#1] SMP PTI
<1>[   22.252640] CPU: 0 PID: 39 Comm: kworker/0:1 Not tainted 4.19.43-microsoft-standard #1
<1>[   22.252780] Workqueue: events p9_read_work
<1>[   22.252800] RIP: 0010:p9_conn_cancel+0x69/0x180
<1>[   22.252803] Code: 56 18 85 d2 0f 85 01 01 00 00 49 8b 46 20 41 89 5e 18 4d 8d 46 20 48 8b 10 48 8d 70 c8 4c 39 c0 48 8d 4a c8 74 33 48 8b 7e 40 <48> 89 7a 08 48 89 17 48 8b 14 24 48 89 04 24 48 89 42 08 48 8d 41
<1>[   22.252808] RSP: 0018:ffffc9000016be20 EFLAGS: 00010283
<1>[   22.252811] RAX: ffff888061aa00c8 RBX: 00000000fffffffb RCX: dead0000000000c8
<1>[   22.252814] RDX: dead000000000100 RSI: ffff888061aa0090 RDI: dead000000000200
<1>[   22.252817] RBP: ffffc9000016be20 R08: ffff888063b42a30 R09: ffff888063b42a5d
<1>[   22.252819] R10: ffffc9000016be28 R11: 0000000000000000 R12: ffff88800fd36000
<1>[   22.252822] R13: 0000000000000000 R14: ffff888063b42a10 R15: 0ffff88806462360
<1>[   22.252826] FS:  0000000000000000(0000) GS:ffff888064600000(0000) knlGS:0000000000000000
<1>[   22.252829] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
<1>[   22.252831] CR2: 00005616fb785fd8 CR3: 000000000220a004 CR4: 00000000000206b0
<1>[   22.252847] Call Trace:
<1>[   22.253023]  p9_read_work+0xc8/0x320
<1>[   22.253102]  process_one_work+0x1ee/0x3e0
<1>[   22.253135]  worker_thread+0x2d/0x3d0
<1>[   22.253139]  ? process_one_work+0x3e0/0x3e0
<1>[   22.253142]  kthread+0x113/0x130
<1>[   22.253174]  ? kthread_create_worker_on_cpu+0x70/0x70
<1>[   22.253236]  ret_from_fork+0x35/0x40
<1>[   22.253254] Modules linked in:
<4>[   22.253287] ---[ end trace 1d50ff7fc82fa884 ]---
<1>[   22.253292] RIP: 0010:p9_conn_cancel+0x69/0x180
<1>[   22.253295] Code: 56 18 85 d2 0f 85 01 01 00 00 49 8b 46 20 41 89 5e 18 4d 8d 46 20 48 8b 10 48 8d 70 c8 4c 39 c0 48 8d 4a c8 74 33 48 8b 7e 40 <48> 89 7a 08 48 89 17 48 8b 14 24 48 89 04 24 48 89 42 08 48 8d 41
<1>[   22.253301] RSP: 0018:ffffc9000016be20 EFLAGS: 00010283
<1>[   22.253304] RAX: ffff888061aa00c8 RBX: 00000000fffffffb RCX: dead0000000000c8
<1>[   22.253307] RDX: dead000000000100 RSI: ffff888061aa0090 RDI: dead000000000200
<1>[   22.253310] RBP: ffffc9000016be20 R08: ffff888063b42a30 R09: ffff888063b42a5d
<1>[   22.253313] R10: ffffc9000016be28 R11: 0000000000000000 R12: ffff88800fd36000
<1>[   22.253316] R13: 0000000000000000 R14: ffff888063b42a10 R15: 0ffff88806462360
<1>[   22.253319] FS:  0000000000000000(0000) GS:ffff888064600000(0000) knlGS:0000000000000000
<1>[   22.253323] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
<1>[   22.253326] CR2: 00005616fb785fd8 CR3: 000000000220a004 CR4: 00000000000206b0
<0>[   22.253354] Kernel panic - not syncing: Fatal exception
<0>[   22.262102] Shutting down cpus with NMI
<0>[   22.262102] Kernel Offset: disabled
```
* What's wrong / what should be happening instead:
should be not crash.

**Comment by benhillis** [comment]
@0xbadfca11 - Thanks.  Looks like another Linux kernel bug in the 9p client.  I will see if @sunilmut has time to take a look.

**Comment by sunilmut** [comment]
This is fixed in upstream Linux kernel by [728356ded](https://github.com/torvalds/linux/commit/728356dedeff8ef999cb436c71333ef4ac51a81c#diff-d9804c9e82370eeb6c891ee6f8f9eae4), but, the fix is not in the WSL2 Linux kernel yet. I will see what we can do to get this in WSL kernel.
